In [90]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [101]:
path = r"C:\Users\madhu\OneDrive\Documents\Naresh IT\Data files\Visadataset.csv"
visa_df = pd.read_csv(path)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [92]:
cat_column = visa_df.select_dtypes(include=object).columns
num_column = visa_df.select_dtypes(exclude=object).columns
num_column

Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object')

**Encoding**

- Encoding means convert categorical column data to Numerical data

- Because Ml models expect the data should be in numerical format

- Ml models developed by Math algorithm

- It is very important to apply encoding

- The following methods are;

    - map
 
    - np.where
 
    - one hot encoding
 
    - Label Encoder

**Map**

- map is  one method to convert categorical values to numerical values

- Take one categorical column

- Get the unique labels first

- Make a dictionary with assign a number to each label

- for example, case_status has two labels

    - Certified
 
    - Denied

- Assign 0 to Certified, 1 to Denied

- Create a dictionary labels as keys, numbers as vakues only

- {'Certified:0,'Denied':1}]



In [5]:
# Step-1: Read the column
# Step-2: Getb the unique values
# Step-3: Make a dictionary
# Step-4: Apply the mapping


In [93]:
visa_df['case_status'].unique()


array(['Denied', 'Certified'], dtype=object)

In [94]:
d = {'Certified':0,'Denied':1}
visa_df['case_status_new'] = visa_df['case_status'].map(d)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  
0          Denied                1  
1       Certified                0  
2          Denied                1  
3          Denied                1  
4       Certified                0  
...           ...              ...  
25475   Certified                0  
25476   Certified                0  
25477   Certified                0  
25478   Certified                0  
25479   Certified                0  

[25480 rows x 13 columns]

In [8]:
visa_df['region_of_employment'].unique()

array(['West', 'Northeast', 'South', 'Midwest', 'Island'], dtype=object)

In [95]:
d = {'West':0,'Northeast':1,'South':2,'Midwest':3,'Island':4}
visa_df['region_of_employment_new'] = visa_df['region_of_employment'].map(d)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  region_of_employment_new  
0          Denied                1                         0  
1       Certified                0                         1  
2          Denied                1                         0  
3          Denied                1                         0  
4       Certified                0                         2  
...           ...              ...                       ...  
25475   Certified                0                         2  
25476   Certified                0                         1  
25477   Certified                0                         2  
25478   Certified                0                         0  
25479   Certified                0                         3  

[25480 rows x 14 columns]

In [96]:

def map_cat_columns(visa_df,cat_column):
    l1=[]
    for i in cat_column:
        unique = visa_df[i].unique()
        l1.append(unique)

    d1_dict = {}
    for j,k in enumerate(cat_column):
        d = {labels:index for index,labels in enumerate(l1[j])} 
        visa_df[k] = visa_df[k].map(d)
        d1_dict[k] = d
    return visa_df, d1_dict[k]   

visa_df, mappings = map_cat_columns(visa_df, cat_column)
visa_df
    

case_id  continent  education_of_employee  has_job_experience  \
0            0          0                      0                   0   
1            1          0                      1                   1   
2            2          0                      2                   0   
3            3          0                      2                   0   
4            4          1                      1                   1   
...        ...        ...                    ...                 ...   
25475    25475          0                      2                   1   
25476    25476          0                      0                   1   
25477    25477          0                      1                   1   
25478    25478          0                      1                   1   
25479    25479          0                      2                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         0         592.2029             0   
1                         1       83425.6500             1   
2                         0      122996.8600             1   
3                         0       83434.0300             1   
4                         2      149907.3900             1   
...                     ...              ...           ...   
25475                     2       77092.5700             1   
25476                     1      279174.7900             1   
25477                     2      146298.8500             1   
25478                     0       86154.7700             1   
25479                     3       70876.9100             1   

       full_time_position  case_status  case_status_new  \
0                       0            0                1   
1                       0            1                0   
2                       0            0                1   
3                       0            0                1   
4                       0            1                0   
...                   ...          ...              ...   
25475                   0            1                0   
25476                   0            1                0   
25477                   1            1                0   
25478                   0            1                0   
25479                   0            1                0   

       region_of_employment_new  
0                             0  
1                             1  
2                             0  
3                             0  
4                             2  
...                         ...  
25475                         2  
25476                         1  
25477                         2  
25478                         0  
25479                         3  

[25480 rows x 14 columns]

In [56]:
# I want to apply the map method for all categorical columns
# Step-1: we need to get unique labels
# Step-2: We need to get the numbers = len unique column
# Step-3: we need to create dictionary


In [66]:
unique_labels = sorted(visa_df['case_status'].unique())
list1 = [i for i in range(len(unique_labels))]
unique_labels,list1
dict1 = dict(zip(unique_labels,list1))
dict1

{'Certified': 0, 'Denied': 1}

In [67]:
{key:value for key,value in zip(unique_labels,list1)}

{'Certified': 0, 'Denied': 1}

In [98]:
unique_labels = sorted(visa_df['case_status'].unique())
list1 = [i for i in range(len(unique_labels))]
unique_labels,list1
dict1 = dict(zip(unique_labels,list1))
dict1
visa_df['case_status'] = visa_df['case_status'].map(dict1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [99]:
for j in cat_column:
    unique_labels = sorted(visa_df[j].unique())
    list1 = [i for i in range(len(unique_labels))]
    unique_labels,list1
    dict1 = dict(zip(unique_labels,list1))
    dict1
    visa_df[j] = visa_df[j].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**Label Encoder**

- LAbelEncoder is ameyhod to convert categorical data into numerical data

- LabelEncodrr is a same way, how we did in the above

- The for loop method we explored

- But the LabelEncoder also do the same

- It is under **scikit-learn** we usually called as **sklearn**

- Under sklearn we have class called as preprocessing

- Because currently we are doing data preprocessing only

- under preprocessing class we have method called as **LabelEncoder**

- sklearn

    - preprocessing
 
        - LabelEncoder
     
- Any sklearn package has 3 steps

    - Step-1: Read the package
 
    - Step-2: Save the package
 
    - Step-3: Apply the fit transform on data

In [76]:
# Step-1: Read the package
from sklearn.preprocessing import LabelEncoder
# Step-2: Save the package
le = LabelEncoder()
# Step-3: Apply the fit transform on data
visa_df['case_status'] = le.fit_transform(visa_df['case_status'])
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [100]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for i in cat_column:
    visa_df[i] = le.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**fit and transform**

- Fit means develop the logic

- Transform means apply the logic

- If we observe in map method first we developed dictionary logic

- Then we apply that dictionary to column

- Developing the dictionary is kind to fit

- Apply the dictionary to column means we are transforming the data cat to num

- whenever we are transforming or changing the data then use **fit transform**

- if we are not changing the data then only **fit**

**np.where**

- Using np.where also we can change the categorical data to numerical data

- But np.where is only for binary condition

- np.where is same as if-else

- if condition is True then we apply True value

- if condition is False then we apply ]the False value

- so np.where use for only binary labels

- A column should have only two unique labels

In [88]:
# value == 'certified' replace with 0
# otherwise 1
# visa_df = pd.read_csv("Naresh IT\Data files\Visadataset.csv")
con = visa_df['case_status'] == 'Certified'
visa_df['case_status'] = np.where(con,0,1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                1  
2                1  
3                1  
4                1  
...            ...  
25475            1  
25476            1  
25477            1  
25478            1  
25479            1  

[25480 rows x 12 columns]

**One hot Encoder**

- One hot means one will be ON, another will be OFF

- ON represents with 1

- OFF represents with 0

- For example, case status has two unique labels

    - certified
    
    - denied

- One hot Encoder creates new column which is equal to number of unique columns

- For example case stauts will create two new extra columns

    - case_status_Certified
 
    - case_status_Denied
 
|case_status|case_status_Certified|case_status_Denied|
|-|-|-|
Denied|0|1|
Certified|1|0|
Denied|0|1|

**pd.get_dummies**

In [109]:
df = pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status'] = visa_df['case_status'].values
df['dummy'] = [i for i in range(len(visa_df))]
df

Certified  Denied case_status  dummy
0              0       1      Denied      0
1              1       0   Certified      1
2              0       1      Denied      2
3              0       1      Denied      3
4              1       0   Certified      4
...          ...     ...         ...    ...
25475          1       0   Certified  25475
25476          1       0   Certified  25476
25477          1       0   Certified  25477
25478          1       0   Certified  25478
25479          1       0   Certified  25479

[25480 rows x 4 columns]

In [154]:
df1 = pd.get_dummies(visa_df['case_status'],dtype='int')
df1

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [155]:
df2 = visa_df[['case_status']]
df2

case_status
0          Denied
1       Certified
2          Denied
3          Denied
4       Certified
...           ...
25475   Certified
25476   Certified
25477   Certified
25478   Certified
25479   Certified

[25480 rows x 1 columns]

**Adding the two dataframes**

- Merge

- Join

- Concat

- Append

In [114]:
pd.get_dummies(visa_df)

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029            True   
1                 2412         2002       83425.6500           False   
2                44444         2008      122996.8600           False   
3                   98         1897       83434.0300           False   
4                 1082         2005      149907.3900           False   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700           False   
25476             3274         2006      279174.7900           False   
25477             1121         1910      146298.8500           False   
25478             1918         1887       86154.7700           False   
25479             3195         1960       70876.9100           False   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0               False           False           False           False   
1                True           False           False           False   
2               False            True           False           False   
3               False           False            True           False   
4               False           False           False            True   
...               ...             ...             ...             ...   
25475           False           False           False           False   
25476           False           False           False           False   
25477           False           False           False           False   
25478           False           False           False           False   
25479           False           False           False           False   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0               False           False  ...                       False   
1               False           False  ...                       False   
2               False           False  ...                       False   
3               False           False  ...                       False   
4               False           False  ...                        True   
...               ...             ...  ...                         ...   
25475           False           False  ...                        True   
25476           False           False  ...                       False   
25477           False           False  ...                        True   
25478           False           False  ...                       False   
25479           False           False  ...                       False   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                           True               True               False   
1                          False              False               False   
2                           True              False               False   
3                           True              False               False   
4                          False              False               False   
...                          ...                ...                 ...   
25475                      False              False               False   
25476                      False              False               False   
25477                      False              False               False   
25478                       True              False               False   
25479                      False              False               False   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                  False              False                 False   
1                  False               True                 False   
2                  False               True                 False   
3                  False               True                 False   
4                  False               True                 False   
...                  ...                ...                  

In [116]:
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

- In One Hot Encoder a new column is created 

- The new column equal to number of unique labels

- Because of case status two extra columns are created

- Because of continenet 6 extra columns are created

- Beacuse of case id 25480 unique ids, 24580 column created

In [118]:
count = 0
for i in cat_column:
    count = count + visa_df[i].nunique()

print(count+len(num_column))

25510


In [120]:
count = 0
for i in cat_column[1:]:
    count = count + visa_df[i].nunique()
print(count)
print(count+len(num_column))

27
30


In [145]:
for i in cat_column[1:]:
    print(visa_df[i].unique(),visa_df[i].nunique())

['Asia' 'Africa' 'North America' 'Europe' 'South America' 'Oceania'] 6
['High School' "Master's" "Bachelor's" 'Doctorate'] 4
['N' 'Y'] 2
['N' 'Y'] 2
['West' 'Northeast' 'South' 'Midwest' 'Island'] 5
['Hour' 'Year' 'Week' 'Month'] 4
['Y' 'N'] 2
['Denied' 'Certified'] 2


**Advantages**

- Using one hot encoder we will get new columns 

- the new column values at a time only one will be **True** another will be **False**

- Which means new columns are **independant to each other**

- Independant means **no relation**

- Means no **Multi Colinearity**

- Multi collinearity is an important property in Machine Learning

- **Machine Learning models expect the input columns should be independant to each other**

- So one hot encoder gives the no multicollineraity

**Disadvantages**

- The disadvantage is it will increase many columns

- In the above example the total columns become 25510

- This sis happend because of case id

- In real time we will drop the id columns

- Eventhough if we drop still we are getting 30 columns

- More columns means, more time to process

- We reuired more infrastructure (CPU-GPU-TPU)

- This is called **Curse of Dimensionality**

In [ ]:
pd.merge(df2,df3,on='case_status')

In [156]:
df1['case_status'] = visa_df['case_status']

In [162]:
merged_df = pd.merge(df, df2, on='case_status')
merged_df

MemoryError: Unable to allocate 2.69 GiB for an array with shape (2, 361217768) and data type int32